pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -82,31 +86,118 @@ impl KeyHistoryRetention {
                     let Value::Image(img) = &logs[0].1 else {
                         unreachable!()
                     };
+                    stat.produce_image_key(img);
                     if let Some(image_writer) = image_writer.as_mut() {
                         image_writer.put_image(key, img.clone(), ctx).await?;
                     } else {
                         delta_writer.push((key, cutoff_lsn, Value::Image(img.clone())));
                     }
                 } else {
                     for (lsn, val) in logs {
+                        stat.produce_key(&val);
                         delta_writer.push((key, lsn, val));
                     }
                 }
                 first_batch = false;
             } else {
                 for (lsn, val) in logs {
+                    stat.produce_key(&val);
                     delta_writer.push((key, lsn, val));
                 }
             }
         }
         let KeyLogAtLsn(above_horizon_logs) = self.above_horizon;
         for (lsn, val) in above_horizon_logs {
+            stat.produce_key(&val);
             delta_writer.push((key, lsn, val));
         }
         Ok(())
     }
 }
 
+#[derive(Debug, Serialize, Default)]
+pub struct CompactionStatistics {
+    num_delta_layer_visited: usize,
```

Comment:
Maybe a `NumSize{ num: u64, size: u64, }` type would be better? Then we'd have half of the number of variables in this struct.

Also, in theory we can process more data than fits in memory so u64 is better than usize (we dont have 32 bit builds but still).

Reply:
resolved in https://github.com/neondatabase/neon/pull/8557/commits/dffe43384bd21dd986c28cafcd8faba243b8179a